### ISSUE TITLE
pdf readonly form fields not visibly filled in when pdf is opened in Acrobat Reader

### ISSUE THREAD

**Issue by smiller** [gettalong/hexapdf#86]
I have a pdf form with read-only fields which I've opened, optimized, and saved in hexapdf (just updated our copy to 0.10, thank you!).  If I open it in a browser or a pdf reader other than Acrobat, the contents of the fields show up (and are read-only) on the page.  

If I open it in Acrobat, in contrast, the form fields show up as empty form fields, until I click into the fields, at which point the expected read-only text shows up.  When I click out of that form field, the text vanishes again.

I see from https://github.com/gettalong/hexapdf/issues/74 that this has been looked at before and the last word appears to be

> the resulting PDF looks fine in okular/evince but still not in Adobe Reader

Is there any further progress on getting it to work with Adobe/Acrobat, or anything I can do within hexapdf at the object level – 

```
pdf.object(3)
#<HexaPDF::Dictionary [3, 0] 
  value={
    :DA=>"/Helv 10 Tf 0g", 
    :F=>4, 
    :FT=>:Tx, 
    :MK=>{}, 
    :P=>
      #<HexaPDF::Type::Page [1, 0] value={:Annots=>#<HexaPDF::Object [2, 0] value=[
        #<HexaPDF::Reference [3, 0]>, 
        #<HexaPDF::Reference [4, 0]>, 
        #<HexaPDF::Reference [26, 0]>, 
        #<HexaPDF::Reference [27, 0]>, 
        #<HexaPDF::Reference [28, 0]>, 
        #<HexaPDF::Reference [29, 0]>
      ]>, 
      :Contents=>[
        #<HexaPDF::Reference [30, 0]>, 
        #<HexaPDF::Reference [31, 0]>, 
        #<HexaPDF::Reference [32, 0]>,
        #<HexaPDF::Reference [33, 0]>, 
        #<HexaPDF::Reference [34, 0]>, 
        #<HexaPDF::Reference [35, 0]>, 
        #<HexaPDF::Reference [36, 0]>, 
        #<HexaPDF::Reference [37, 0]>
      ], 
      :CropBox=>#<HexaPDF::Rectangle [0, 0] value=[0, 0, 612, 792]>, 
      :MediaBox=>#<HexaPDF::Rectangle [0, 0] value=[0, 0, 612, 792]>, 
      :Parent=>#<HexaPDF::Type::PageTreeNode [40, 0] value={:Type=>:Pages, :Kids=>[#<HexaPDF::Reference [1, 0]>, #<HexaPDF::Reference [7, 0]>], :Count=>2}>, 
      :Resources=>#<HexaPDF::Type::Resources [38, 0] value={:ColorSpace=>{:Cs1=>#<HexaPDF::Reference [14, 0]>}, :Font=>{:TT2=>#<HexaPDF::Reference [18, 0]>, :TT4=>#<HexaPDF::Reference [22, 0]>}, 
      :ProcSet=>[:PDF, :Text]}>, 
      :Type=>:Page}>, 
    :Rect=>[101.434, 682.363, 251.434, 697.022], 
    :Subtype=>:Widget, 
    :T=>"Closing~DocsDate", 
    :V=>"2019-09-09", 
    :Ff=>1
  }
>
```

– to make it show up in Acrobat not as an (apparently) empty widget but as the text value (`:V=>"2019-09-09"`) that I can only see now by clicking inside the widget?

**Comment by smiller** [comment]
Here also is the `hexapdf inspect` output for that object, if that is more useful/digestible:

```
cmd> 3
<<
  /DA (/Helv 10 Tf 0 g)
  /F 4
  /FT /Tx
  /MK <<
  >>
  /P 1 0 R
  /Rect [101.434 682.363 251.434 697.022 ]
  /Subtype /Widget
  /T (Closing~DocsDate)
  /V (2019-09-09)
  /Ff 1
>>
```

**Comment by smiller** [comment]
The example was simplified to focus on the problem: there are 10-20 pdf forms.  I'm using another library to find all the fields and set the `field[:V]` values (and sometimes change the font size in `field[:DA]` when auto-resizing-to-fit doesn't seem to work), and then I use HexaPDF to combine all the forms into a single pdf file, and I ran various configurations of optimize – currently

```
package.task(:optimize, compact: true,
                                object_streams: :delete,
                                compress_pages: false)
```

though I also tried all the other `object_streams:` and `compress_pages:` options – to see if any of them made the fields visible in Adobe.

diff outputs to follow shortly, thank you!

**Comment by smiller** [comment]
The whole file diff is 260 lines – (but let me know if that's still useful and I can totally gist it and link to it) – but the `Closing~DocsDate` element is 

```
44 0 obj
<<
  /DA (/Helv 10 Tf 0 g)
  /F 4
  /FT /Tx
  /MK <<
  >>
  /P 26 0 R
  /Rect [101.434 682.363 251.434 697.022 ]
  /Subtype /Widget
  /T (Closing~DocsDate)
  /Type /Annot
  /V (2019-09-09)
  /Ff 1
>>
```

before optimize, and 

```
3 0 obj
<<
  /DA (/Helv 10 Tf 0 g)
  /F 4
  /FT /Tx
  /MK <<
  >>
  /P 1 0 R
  /Rect [101.434 682.363 251.434 697.022 ]
  /Subtype /Widget
  /T (Closing~DocsDate)
  /V (2019-09-09)
  /Ff 1
>>
```

after optimize.

Or, much lower object number and optimize has taken out the `/Type /Annot` line.  (In either case, it initially shows up as a blank form field in Acrobat.)

**Comment by smiller** [comment]
At one point I was trying to set `field[:Subtype]` to something other than `:Widget`, on the grounds that it won't print a form widget if that's not what the pdf says it is, but I couldn't find the subtype that was just “transparently display the object value”.  (`:Text` didn't.)

**Comment by gettalong** [comment]
Acroform only uses annotations of subtype `:Widget`, all other annotation subtypes are really just annotations.

If you could provide the whole diff, it may help - thank you! The missing `/Type /Annot` is okay because that is the default value for this field, lower object numbers are also okay and expected.

Alas, Adobe Acrobat sometimes does things **exactly** by the book (=spec) without wriggle room and other times does things **exactly contrary** to the spec or at least in a way not described by the spec... I will see if I find anything useful in the specification.

**Comment by smiller** [comment]
Pre-optimized `hexapdf inspect` is at: https://gist.github.com/smiller/05511ccd4c42dd3906f0356be9318847
Post-optimized `hexapdf inspect` is at: https://gist.github.com/smiller/d9c8010901924b66a8a19aa3b9ec32f9
Diff is at: https://gist.github.com/smiller/90d5471ad57e4ebde2ad44749fda82e3

This looks so unwieldy though that I'm tempted to do a version that does the field-finding and `field[:V]`-setting in HexaPDF just to remove the excess differences.

**Comment by smiller** [comment]
That made less difference to the unwieldy-ness of the diff than I hoped, but fresh set where HexaPDF is used to set `field[:V]` and `field[:Ff]` (for read-only):

Pre-optimized: https://gist.github.com/smiller/91c32a8803315e25fc1bdf85da8c2273
Post-optimized: https://gist.github.com/smiller/cffe15cf535c55c2e748e67000d0264c
Diff: https://gist.github.com/smiller/9a7903213e6359dbc9795e5ec5194e80

**Comment by smiller** [comment]
(Also, thank you very much for digging in, and if I can be providing this data in any form that is more useful for you, please let me know.)

**Comment by gettalong** [comment]
Thanks for all the data!

One thing I noticed straight ahead is that the main `/AcroForm` key is missing from the post-optimized PDF, in both variants. Is this something you removed manually or was this done by HexaPDF? If it was the latter, it would be helpful if you provide a sample PDF along with code to reproduce the post-optimized file.

Another question: Does the pre-optimized PDF that was produced by HexaPDF work correctly in Adobe Acrobat?

**Comment by smiller** [comment]
I'm checking with the business about sharing the files, since base file is a business template – if they have a problem I'll create a dummy file that displays the same problem and post that instead – but, meanwhile:

1. I didn't remove anything manually.  The code I used to produce the two files that use HexaPDF to set the field[:V] value is

```
template = "1952"
template_name = "#{DocGen::Queries::Roots.templates}/template_#{template}.pdf"

pdf = HexaPDF::Document.open(template_name)

field = pdf.object(44) # found by consoling; presumably there's some more efficient way to get the field with [:T] of "Closing~DocsDate"
field[:V] = "2019-09-09"
field[:Ff] ||= 0
field[:Ff] |= Flags::READONLY

job_id = "test"
package_name = "#{DocGen::Queries::Roots.packages}/hex-package-#{job_id}-#{template}.pdf"
pdf.write(package_name)

optimized_package_name = "#{DocGen::Queries::Roots.packages}/hex-package-#{job_id}-#{template}-optimized.pdf"

package = HexaPDF::Document.new
# in the real code, of course, this is more like pdfs.each { |pdf| … }
pdf = HexaPDF::Document.open(package_name)
pdf.pages.each { |page| package.pages << package.import(page) }
package.task(:optimize, compact: true,
                        object_streams: :delete)
package.write(optimized_package_name)
```

2. And no, the pre-optimized file displays the same behaviour in Acrobat (that is, Closing~DocsDate field appears to be a blank form field, which you can click into to discover the date set.)

**Comment by smiller** [comment]
I was able to reproduce the problem with a much simpler form.  See now https://github.com/smiller/pdf-form-test, or, more immediately:

form template (not yet filled in):

- https://github.com/smiller/pdf-form-test/blob/master/SimpleForm.pdf

script used to build filled in forms:

- https://github.com/smiller/pdf-form-test/blob/master/fill_in_form.rb

filled in form, not yet optimized:

- https://github.com/smiller/pdf-form-test/blob/master/FilledInForm.pdf
- https://github.com/smiller/pdf-form-test/blob/master/hexapdf-ins-FilledInForm

filled in form, optimized:

- https://github.com/smiller/pdf-form-test/blob/master/FilledInFormOptimized.pdf
- https://github.com/smiller/pdf-form-test/blob/master/hexapdf-ins-FilledInFormOptimized


`FilledInForm.pdf` is the pre-optimized but filled in form (hexapdf ins at `hexapdf-ins-FilledInForm`)
`FilledInFormOptimized.pdf` is the optimized filled in form (hexapdf ins at `hexapdf-ins-FilledInFormOptimized`.

**Comment by gettalong** [comment]
Tl;dr: Solution found, see second heading below!

## About the optmized version

The `./FilledInFormOptimized.pdf` file doesn't display the filled in date, at least not in Evince, Okular or acroread on Linux. The possible reason for this is that the `/AcroForm` entry is missing because of the way the file is built. The original file is not used but its pages are imported into a new PDF which means that the `/AcroForm` key is **not imported**. If you do the following, it will show in Evince and Okular but still not in acroread:

~~~ ruby
pdf = HexaPDF::Document.open(filled_in_name)
pdf.task(:optimize, compact: true,
                        object_streams: :delete)

pdf.write("FilledInFormOptimized2.pdf")
~~~

However, this approach will only work on a single file. If you need to combine multiple files, you would need to correctly merge the `/AcroForm` entry.

## Back to the problem and its solutiong

Comparing the original `SimpleForm.pdf` with `FilledInForm.pdf` leads to marginal changes, namely only those that were scripted (`/Ff` and `/V`) and some normal expected changes by HexaPDF in the root and info dictionaries.

So why does Acrobat not show the value like other viewers? I found [this StackOverflow question](https://stackoverflow.com/questions/38915591/adobe-pdf-forms-text-field-displays-value-only-when-clicked-on-it) which guesses that Acrobat does only generate appearance streams if explicitly instructed to do so by setting the `/NeedAppearances` entry in the `/AcroForm` dictionary. Put the following after opening the original file in `fill_in_form.pdf`:

~~~
pdf.catalog[:AcroForm][:NeedAppearances] = true
~~~

Then run the script and acroread will show the text! :tada: I don't have Windows available so please test on a recent version of Acrobat if it also shows the text.

It seems that most viewers do the sensible thing automatically, i.e. generate an appearance stream if there is a value but no appearance stream. Acrobat, on the other hand, seems to strictly follow the spec here and only generates the appearance stream for existing values if the above mentioned key is set.

Thank you for helping to figure this one out!

**Comment by smiller** [comment]
@gettalong Is there a missing line in your first code block to actually add the `/Acroform` entry?

What you've got

```
pdf = HexaPDF::Document.open(filled_in_name)
pdf.task(:optimize, compact: true,
                        object_streams: :delete)

pdf.write("FilledInFormOptimized2.pdf")
```

appears to be the same as what was in the original script, and when I add 

```
pdf.catalog[:Acroform][:NeedAppearances] = true
```

as line 19 and try to run the script I get

```
> ruby fill_in_form.rb
Traceback (most recent call last):
fill_in_form.rb:19:in `<main>': undefined method `[]=' for nil:NilClass (NoMethodError)
```

**Comment by smiller** [comment]
@gettalong Hmm.  Using Adobe Acrobat Reader DC or Adobe Acrobat Pro DC on a Mac, I'm still seeing the field-is-only-visible-when-you-click-on-it behaviour even after adding the

```
pdf.catalog[:AcroForm][:NeedAppearances] = true
```

line.  However, the main problem was the Windows version of acroread, and I do have access to Windows, so checking that now.

**Comment by smiller** [comment]
@gettalong On a Windows box with Adobe Acrobat Reader DC and the line 

```
pdf.catalog[:AcroForm][:NeedAppearances] = true
```

added to `fill_in_forms.rb`, I'm still seeing the previous behaviour, that is, open the revised FillInFormOptimized.pdf in Adobe Acrobat Reader DC and it initially just shows `Date: ___________________` and only shows the actual filled-in date when you click on the line where the date ought to be.

I've created a new branch of pdf-form-test, https://github.com/smiller/pdf-form-test/tree/with-NeedAppearances-fix, with the extra `pdf.catalog[:AcroForm][:NeedAppearances] = true` line and the regenerated FilledInFormOptimized.pdf file – could you please have a look to see if there's anything there I'm missing that I might still need to get this to work?

**Comment by gettalong** [comment]
Move [this line](https://github.com/smiller/pdf-form-test/blob/with-NeedAppearances-fix/fill_in_form.rb#L19) directly after line 10, then 'FilledInForm.pdf` should work. The optimized version won't work as explained above - if you use [my code](https://github.com/gettalong/hexapdf/issues/86#issuecomment-544110920) it should work.

**Comment by smiller** [comment]
Back and trying this out now – thank you!

**Comment by smiller** [comment]
Excellent.  Thank you very much!  Can confirm, that the now-corrected code, as seen in https://github.com/smiller/pdf-form-test/tree/with-NeedAppearances-fix-try-2, displays the correct behaviour for a single file on both Mac and Windows.  Hurrah!

Just so we've got all the code in one place, with this, both FilledInForm.pdf and FilledInFormOptimized.pdf display the field properly when opened in Acrobat:

```
# Get empty form
template_name = "./SimpleForm.pdf"
pdf = HexaPDF::Document.open(template_name)

# Set field value in form
field = pdf.object(22)
field[:V] = "2019-09-09"
field[:Ff] ||= 0
field[:Ff] |= (1 << 0) # make it readonly

# Set [:NeedAppearances], required for Acrobat
pdf.catalog[:AcroForm][:NeedAppearances] = true

# Save form with field and NeedAppearances
filled_in_name = "./FilledInForm.pdf" 
pdf.write(filled_in_name)

# Save optimized version
optimized_name = "./FilledInFormOptimized.pdf" 
pdf = HexaPDF::Document.open(filled_in_name)
pdf.task(:optimize, compact: true,
                    object_streams: :delete)
pdf.write(optimized_name)
```

For the single-file case this is now therefore fixed, and I really appreciate the rapid response.

Unfortunately I've still got the combining-multiple-files case, which means I'll be exploring

> However, this approach will only work on a single file. If you need to combine multiple files, you would need to correctly merge the /AcroForm entry.

next.  Unless I can create all the files, write them with `[:NeedsAppearances]`, and then after that combine them?  That's the first thing I'm going to try, anyway.  Will report progress.

**Comment by smiller** [comment]
Can confirm (see https://github.com/smiller/pdf-form-test/tree/with-NeedAppearances-fix-multiple-files) that just creating multiple files with the `[:NeedAppearances]` fix and combining the optimized files doesn't create a combined file that displays correctly in Acrobat (that is, in CombinedForms.pdf we're back to the open in Acrobat and it shows a form field instead of the field value, and only shows the value if you click into the field).

So 

> However, this approach will only work on a single file. If you need to combine multiple files, you would need to correctly merge the /AcroForm entry.

is still an extant problem for the combining-multiple-files case.

But since the solution provided absolutely works in the single-file case, I'd be happy to agree that this issue is resolved and open another issue for the multiple-files case if that would be better.

And I'll continue to poke at this regardless.  Thanks again for your help.

**Comment by gettalong** [comment]
No problem. Try the following code (you need to adapt it a bit) with multiple files:

~~~ ruby
package = HexaPDF::Document.new
pdf = HexaPDF::Document.open(filled_in_name)
pdf.pages.each { |page| package.pages << package.import(page) }
(package.catalog[:AcroForm] = {})[:NeedAppearances] = true
package.task(:optimize, compact: true,
                        object_streams: :delete)
~~~

I just added the needed entry, leaving out all other entries that we would possible need to fill in to comply to the spec. Seems to work in the above case, although Acroread doesn't recognize the form fields anymore. Maybe will also work with multiple files.

**Comment by smiller** [comment]
Ah.  Great.  Trying now, thank you!

**Comment by gettalong** [comment]
One more thing: I'm currently starting to implement acroform support. So this will hopefully work better and more correctly in the near future.

**Comment by smiller** [comment]
Yes, that does the trick, thank you.  Can confirm (by https://github.com/smiller/pdf-form-test/tree/with-NeedAppearances-fix-multiple-files-try-2) that with that extra line ensuring `[:NeedAppearances]` is also in the file into which we're combining files, the `CombinedFiles.pdf` opened in Acrobat on Mac or Windows shows the contents of the form fields.  (Which is all I need it to do, since once they're filled in they should be read-only and I no longer care about the form fields.)

Thank you very much indeed for all your help, this totally counts as closable as far as I'm concerned.

I do have one more question, since it looks like I might now also be using HexaPDF for filling in the form fields to set the initial `[:NeedsAppearance]` on the files before combining them.  Starting from

```
pdf = HexaPDF::Document.open("SimpleForm1.pdf")
```

where SimpleForm1.pdf has a field called `Closing~DocsDate`, is there anything like 

```
field = pdf.get_field("Closing~DocsDate")
```

or, as inefficient pseudocode,

```
pdf.get_all_fields.select { |f| f[:T] == "#{desired_field_name}" }
```

?  Given `get_all_fields`, I can inefficiently get the rest of the way there, but that's the missing piece.  And of course if there's a more efficient / elegant way of getting there in the first place, even better.

But, regardless, and again, thank you very much indeed for all your help.

**Comment by gettalong** [comment]
I would have suggested something like you already worked out. As I have written I'm in the process of implementing proper support for AcroForm. Details are not fixed yet but most probably there will be a `document.acroform` method for getting the main acroform object and then various convenience methods like `fields` which would return all fields or `fields(page_number)` returning all fields from that page.

**Comment by gettalong** [comment]
@smiller Just a heads up: AcroForm support is coming with the next release tentatively scheduled for end of June. This will allow you to easily find fields, create appearance streams and update values. I have pushed the changes to the [devel](https://github.com/gettalong/hexapdf/tree/devel/lib/hexapdf/type/acro_form) branch on the repo.

It would be great if you could have a look at the code and let me know if it would simplify your use cases. And would also be very grateful for pointers on what is missing for you.

E.g. [`acro_form.find_root_fields!`](https://github.com/gettalong/hexapdf/blob/devel/lib/hexapdf/type/acro_form/form.rb#L100) could be run after combining files to properly store the list of fields in the main AcroForm object (which can be created by `document.acro_form(create: true)`. And [`acro_form.field_by_name(name)`](https://github.com/gettalong/hexapdf/blob/devel/lib/hexapdf/type/acro_form/form.rb#L123) helps with finding a certain field.

**Comment by smiller** [comment]
Hey @gettalong , this looks like it would absolutely let us remove some of the custom scaffolding code that I built to find and identify fields on pdf forms.  I'm no longer directly involved with that codebase – else I'd already have sent you a report – but I've got in touch with the person who inherited it and asked if he'd like to do a remote pairing session to walk through the changes we can now make, and I'll get back to you when I have details.  (And thanks again very much for your generous help last year when I was struggling with this!)

**Comment by gettalong** [comment]
No problem, glad I could help! There are still some more changes lined up, e.g. support for check boxes and radio buttons. They will work similarly to the existing code.

**Comment by gettalong** [comment]
@smiller The new release of HexaPDF is out with support for interactive forms. If you (or the person in charge of the code base) have any problems with it, please open a new issue
